src/coreclr/src/vm/callcounting.cpp

```diff
@@ -448,7 +448,7 @@ bool CallCountingManager::IsCallCountingEnabled(NativeCodeVersion codeVersion)
     CONTRACTL_END;
 
     _ASSERTE(!codeVersion.IsNull());
-    _ASSERTE(codeVersion.IsDefaultVersion());
+    _ASSERTE(codeVersion.IsDefaultVersion() || codeVersion.GetILCodeVersionId() != 0);
```

Comment:
This function shouldn't be called for non-default native code versions. It effectively is a storage location just for the default code version that affects the tier for the default code version. Non-default code versions have a storage location for the tier. It looks like it this caller would be the issue (there are only two callers): https://github.com/dotnet/runtime/blob/2b27323e4aa61f540b8c6600f4add5b29860a0a3/src/coreclr/src/vm/prestub.cpp#L1051-L1056. That assert is actually unnecessary since the assert above it checks for tier 0, which implies what is intended to be verified. Could remove that assert instead (including the call to this function) and the only other call to this function should be fine. Do you have the stack trace of the failure to confirm? If you can take care of it then great, otherwise I'm working on a fix and I can include this fix in that, let me know.

Reply:
The issue I saw is that non-default IL versions will have a tier 0 that returns false for NativeCodeVersion::IsDefaultVersion. I haven't seen any evidence that it's called for tier 1 methods

Reply:
I don't have the stack trace any more, I can get it but it would have to be later today

Reply:
Yea makes sense. It would be great if you could confirm that as the cause, I'll go ahead and remove the assert in my change anyway. Also can you point me to the test that is failing?

Reply:
I'm removing the other assertion I mentioned above in https://github.com/dotnet/runtime/pull/33512. I have confirmed that it is the cause of the failure and verified the fix. @davmason, you can go ahead and revert this part of the change. Thanks!